### ISSUE TITLE
Scheduler always steps when training with gradient accumulation

### ISSUE THREAD

**Issue by markovalexander** [huggingface/accelerate#963]
### System Info

```Shell
- Accelerate version: 0.15.0
- Platform: macOS-10.16-x86_64-i386-64bit
- Python version: 3.8.11
- Numpy version: 1.23.0
- PyTorch version (GPU?): 1.9.0 (False)
- Accelerate default config:
        Not found
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

When training a model with gradient accumulation scheduler modifies learning rate at every step, even when `optimiser.step` was not performed.

To reproduce:
1. Add debugging prints 🙂  [here](https://github.com/huggingface/accelerate/blob/9b7ef9679f3a3d1c3e579562568f6fc637e62091/src/accelerate/optimizer.py#L129) and [here](https://github.com/huggingface/accelerate/blob/9b7ef9679f3a3d1c3e579562568f6fc637e62091/src/accelerate/scheduler.py#L53) after `.step()` calls to track them
2. Run the following ([pastebin](https://pastebin.com/MZimGGwc)) script with `accelerate launch test.py`
```
from accelerate import Accelerator
from transformers import (
    get_cosine_schedule_with_warmup,
)
import torch
 
accelerator = Accelerator(
    device_placement=True,
    gradient_accumulation_steps=4,
    split_batches=True,
)
 
model = torch.nn.Linear(1, 1)
optimizer = torch.optim.AdamW(model.parameters())
scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=10, num_training_steps=50)
model, optimizer, scheduler = accelerator.prepare(model, optimizer, scheduler)
 
for i in range(50):
    with accelerator.accumulate(model):
        print('----------------------------- batch -----------------------------')
        optimizer.zero_grad()
        loss = model(torch.randn(10, 1))
        accelerator.backward(loss.mean())
 
        optimizer.step()
        scheduler.step()
```
Output: 
![telegram-cloud-photo-size-2-5325883662245414381-y](https://user-images.githubusercontent.com/22663468/211563928-bdb5093a-7817-4e46-a499-801f618756fb.jpg)

### Expected behavior

`scheduler.step()` is called only after `optimizer.step()` every `gradient_accumulation_steps` iteration.

The possible solution may be to modify `optimizer.step_was_skipped` property to look somehow like this:

```
    def step_was_skipped(self):
        """Whether or not the optimizer step was skipped."""
        return self._is_overflow or (not self.gradient_state.sync_gradients)
```

I can make a PR

**Comment by sgugger** [comment]
I'm not too sure where the bug lies here: you print the statement in the beginning of the step method of the scheduler but after the test for gradient accumulation in the optimizer step. You should take into account the early return [performed here](https://github.com/huggingface/accelerate/blob/9b7ef9679f3a3d1c3e579562568f6fc637e62091/src/accelerate/scheduler.py#L62) which tests whether the optimizer was not stepped.

**Comment by markovalexander** [comment]
@sgugger  no no, I just gave a link to step method. I added `print` after each line where `scheduler.step()` is called

**Comment by markovalexander** [comment]
@sgugger Actually I think that early return is performed only when overflow encountered (and I believe it is not even possible for fp32 and bf16): [here](https://github.com/huggingface/accelerate/blob/9b7ef9679f3a3d1c3e579562568f6fc637e62091/src/accelerate/optimizer.py#L159) and [here](https://github.com/huggingface/accelerate/blob/9b7ef9679f3a3d1c3e579562568f6fc637e62091/src/accelerate/optimizer.py#L138).

**Comment by pacman100** [comment]
Hello @markovalexander, when using gradient accumulation, the `training_steps = len(train_dataloader) * num_epochs/gradient_accumulation_steps`. Now, if you provide scheduler `num_training_steps =len(train_dataloader) * num_epochs`  you are accounting for the gradient accumulation steps by having scheduler always step, i.e., scheduler steps gradient_accumulation_steps times for every optimizer step: https://github.com/huggingface/accelerate/blob/main/examples/by_feature/gradient_accumulation.py#L14

**Comment by markovalexander** [comment]
Hi @pacman100  !

> i.e., scheduler steps gradient_accumulation_steps times for every optimizer step

yeah, that is the issue I am talking about, why do you think this is a correct behaviour?

**Comment by pacman100** [comment]
Hello @markovalexander, let's take an example:
len(train_dataloader) = 100
num_epochs = 10
gradient_accumulation_steps = 4
lr = 5e-5 
Scheduler is linear scheduler with no warmup steps for simplicity.

Now, 
training_steps = 100*10/4 = 250. So, scheduler should step 250 times to go from 5e-5 to 0.
We set scheduler's `num_training_steps` = 100*10 =1000 and have it step always. Now, after 4 scheduler steps the lr will be 
4.98e-5 (i.e., 5e-5 * (1-4/1000)). It is same as  a scheduler having `num_training_steps=250` and stepping only when optimizer steps wherein after 1 sceduler step the lr will be again 4.98e-5 (i.e., 5e-5 * (1-1/250))

**Comment by markovalexander** [comment]
@pacman100 Ah, I see. It makes sense, and the [documentation](https://huggingface.co/docs/accelerate/v0.14.0/en/package_reference/torch_wrappers#accelerate.scheduler.AcceleratedScheduler) says that we should not do it so it sounds a little bit confusing in case we parametrize the total number of training steps we want to make.

In this case we have to adjust scheduler's length contrary to the documentation, but I think it is just another usecase and we should call `scheduler.step()` inside `if accelerator.sync_gradient` statement.

**Comment by sgugger** [comment]
Yes you should not have to adapt the length of your scheduler, so it looks like there is a bug to fix here. @muellerzr or I will have a look next week :-)

**Comment by yuvalkirstain** [comment]
Is fixing the bug anything more than adding:
```python
if self.gradient_state.sync_gradients:
```
at the start of the `step` method in `acceleratedScheduler`?
If it is enough, I can create a PR :)

**Comment by markovalexander** [comment]
> Hello @markovalexander, let's take an example: len(train_dataloader) = 100 num_epochs = 10 gradient_accumulation_steps = 4 lr = 5e-5 Scheduler is linear scheduler with no warmup steps for simplicity.
> 
> Now, training_steps = 100*10/4 = 250. So, scheduler should step 250 times to go from 5e-5 to 0. We set scheduler's `num_training_steps= 100*10 =1000` and have it step always. Now, after 4 scheduler steps the lr will be 4.98e-5 (i.e., 5e-5 * (1-4/1000)). It is same as a scheduler having `num_training_steps=250` and stepping only when optimizer steps wherein after 1 sceduler step the lr will be again 4.98e-5 (i.e., 5e-5 * (1-1/250))

By the way, I think that this approach is completely incorrect when we use schedulers with warmup since we simply skip some (or even all) warmup steps

**Comment by pacman100** [comment]
> By the way, I think that this approach is completely incorrect when we use schedulers with warmup since we simply skip some (or even all) warmup steps

How so?

Continuing with the example

len(train_dataloader) = 100
num_epochs = 10
gradient_accumulation_steps = 4
lr = 5e-5
num_warum_steps = 10% of all steps

Now,
training_steps = 100*10/4 = 250, num_warum_steps=25. So, lr goes from 0 to 5e-5 in 25 steps and then decreases from 26th step to 0 in 225 steps.

We set scheduler's num_training_steps= 100*10 =1000 and num_warum_steps=100. Now, after 4 scheduler steps the lr will be 2e-6 (i.e., 5e-5 * (4/100)). It is same as a scheduler having num_training_steps=250 and stepping only when optimizer steps wherein after 1 sceduler step the lr will be again 2e-6 (i.e., 5e-5 * (1/25))

**Comment by markovalexander** [comment]
> > By the way, I think that this approach is completely incorrect when we use schedulers with warmup since we simply skip some (or even all) warmup steps
> 
> How so?
> 
> Continuing with the example
> 
> len(train_dataloader) = 100 num_epochs = 10 gradient_accumulation_steps = 4 lr = 5e-5 num_warum_steps = 10% of all steps
> 
> Now, training_steps = 100*10/4 = 250, num_warum_steps=25 (10%). So, lr goes from 0 to 5e-5 in 25 steps and then decreases from 26th step to 0 in 225 steps.
> 
> We set scheduler's num_training_steps= 100*10 =1000 and num_warum_steps=100 (10%). Now, after 4 scheduler steps the lr will be 2e-6 (i.e., 5e-5 * (4/100)). It is same as a scheduler having num_training_steps=250 and stepping only when optimizer steps wherein after 1 sceduler step the lr will be again 2e-6 (i.e., 5e-5 * (1/25))

Well, I think it is more common (imho) to pass a fixed number of warmup steps, i.e 

num_warmup_steps = 10
and not
num_warmup_steps = len(train_loader) * warmup_fraction

and the exact number of warmup_steps would be floor(num_warmup_steps / gradient_accumulation_steps), that might be even equal to 0

**Comment by sgugger** [comment]
> Is fixing the bug anything more than adding:
> 
> ```python
> if self.gradient_state.sync_gradients:
> ```
> 
> at the start of the `step` method in `acceleratedScheduler`? If it is enough, I can create a PR :)

I think this is the right fix, so please go ahead @yuvalkirstain
